File: Sem3/Advanced Methods of Programming/MyInterpreter - JavaFX Graphical Application/project/src/domain/expression/ArithmeticExpression.java
package domain.expression;


import domain.adt.MyDictionaryInterface;


public class ArithmeticExpression extends Expression
{
    private Expression e1;
    private Expression e2;
    private char operator;


    public ArithmeticExpression(Expression e1, Expression e2, char operator) throws ExpressionException
    {
        if (operator != '+' && operator != '-' && operator != '*' && operator != '/' && operator != '%')
            throw new ExpressionException("Invalid operator.");
        this.e1 = e1;
        this.e2 = e2;
        this.operator = operator;
    }


    @Override
    public int eval(MyDictionaryInterface<String, Integer> symbolTable,
                    MyDictionaryInterface<Integer, Integer> heapTable) throws ExpressionException
    {
        if (operator == '+')
            return e1.eval(symbolTable, heapTable) + e2.eval(symbolTable, heapTable);
        else if (operator == '-')
            return e1.eval(symbolTable, heapTable) - e2.eval(symbolTable, heapTable);
        else if (operator == '*')
            return e1.eval(symbolTable, heapTable) * e2.eval(symbolTable, heapTable);
        else if (e2.eval(symbolTable, heapTable) == 0)
            throw new ExpressionException("Divide by 0.");
        else if (operator == '/')
            return e1.eval(symbolTable, heapTable) / e2.eval(symbolTable, heapTable);
        else if (operator == '%')
            return e1.eval(symbolTable, heapTable) % e2.eval(symbolTable, heapTable);
        else
            return 0;
    }


    @Override
    public String toString()
    {
        return "(" + e1.toString() + operator + e2.toString() + ")";
    }
}

File: Sem3/Advanced Methods of Programming/MyInterpreter - CSharp Console Application/MyInterpreter CSharp/domain/expression/ArithmeticExpression.cs
using System.Collections.Generic;
using MyInterpreter_CSharp.domain.adt;

namespace MyInterpreter_CSharp.domain.expression
{
    public class ArithmeticExpression : Expression
    {
        private Expression _e1;
        private Expression _e2;
        private char _op;

        public ArithmeticExpression(Expression e1, Expression e2, char op)
        {
            if (op != '+' && op != '-' && op != '*' && op != '/' && op != '%')
                throw new ExpressionException("Invalid operator.");
            _e1 = e1;
            _e2 = e2;
            _op = op;
        }


        public override int Eval(IMyDictionary<string, int> symbolTable)
        {
            if (_op == '+')
                return _e1.Eval(symbolTable) + _e2.Eval(symbolTable);
            if (_op == '-')
                return _e1.Eval(symbolTable) - _e2.Eval(symbolTable);
            if (_op == '*')
                return _e1.Eval(symbolTable) * _e2.Eval(symbolTable);
            if (_e2.Eval(symbolTable) == 0)
                throw new ExpressionException("Divide by 0.");
            if (_op == '/')
                return _e1.Eval(symbolTable) / _e2.Eval(symbolTable);
            if (_op == '/')
                return _e1.Eval(symbolTable) % _e2.Eval(symbolTable);
            return 0;
        }

        public override string ToString()
        {
            return _e1 + "" + _op + "" + _e2;
        }
    }
}